### ISSUE TITLE
Shutdown halts while shutting down subsystems

### ISSUE THREAD

**Issue by mayant15** [MovingBlocks/Terasology#3738]
Game shutdown halts while shutting down engine subsystems. The problem occurs at shutdown and not pre-shutdown, maybe somewhere in this loop [here](https://github.com/MovingBlocks/Terasology/blob/70bf0aa4a21fde2c15bd67bd7e76eb0274106218/engine/src/main/java/org/terasology/engine/TerasologyEngine.java#L501).

### How to reproduce

* Start the game
* Exit the game
* The window closes but shutdown halts

### Logs

The log after when the game was shutdown:
```
04:50:05.439 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
04:50:05.462 [Thread-5] INFO  o.t.e.s.rpc.DiscordRPCSubSystem - Discord RPC >> Disconnected!
```
It stops after disconnecting Discord RPC.

A waiting thread:
```
"pool-3-thread-12@4183" prio=5 tid=0x34 nid=NA waiting
  java.lang.Thread.State: WAITING
	  at sun.misc.Unsafe.park(Unsafe.java:-1)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	  at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	  at java.lang.Thread.run(Thread.java:748)
```
Full thread dump can be found [here](https://pastebin.com/cwg0WTm6)

### Details
Commit: 8edc61254
OS: Windows 10 (Version 1903 Build 18362.10012)
JDK: 1.8.0_211

**Comment by Cervator** [comment]
Can also confirm this seems to happen even solely with SampleBuilderGameplay which is the absolute minimal amount of stuff to run the engine with. So it doesn't seem tied to any module content.

I tried to look at my own thread dumps, but the target a bunch of threads are waiting on doesn't appear to exist in the output ... which is either weird to me or a symptom of a wait target having vanished without releasing stuff waiting on it? I don't know threading to that level :-)

**Comment by Cervator** [comment]
Looks like we fixed this a while ago - maybe with #3804 as noted but unsure, so counting it as complete with the current milestone :-)

There are still some other cases that can leave the game orphaned, but we're slowly fixing the leaks. Thanks for the report @mayant15 and likely fix @DarkWeird 👍
